packages/core/src/application.ts

```diff
@@ -315,12 +366,50 @@ export class Application extends Context implements LifeCycleObserver {
     this.add(binding);
     return binding;
   }
+
+  /**
+   * Set up signals that are captured to shutdown the application
+   * @param signals - An array of signals to be trapped
+   * @param gracePeriod - A grace period in ms before forced exit
+   */
+  protected setupShutdown(signals: NodeJS.Signals[], gracePeriod?: number) {
+    const cleanup = async (signal: string) => {
+      const kill = () => {
+        // eslint-disable-next-line @typescript-eslint/no-misused-promises
+        signals.forEach(sig => process.removeListener(sig, cleanup));
+        process.kill(process.pid, signal);
```

Comment:
IIUC, this code is trying to kill the process by sending it the same signal that we have already intercepted. Is this the right usage of `process.on` feature? I mean aren't we kind of creating an infinite loop, where the signal handler send the same signal again?

@sam-github would you mind lending us our expertise? We are trying to intercept signals like KILL so that we can perform application-specific cleanup before exiting the process.

Reply:
No, it's intentional. The first `kill` is trapped to run `stop` as the `_isShuttingDown` flag is `false`. The 2nd `kill` proceeds with routine Node.js default logic to shut down the process.